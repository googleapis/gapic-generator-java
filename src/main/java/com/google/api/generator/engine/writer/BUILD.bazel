package(default_visibility = ["//visibility:public"])

filegroup(
    name = "writer_files",
    srcs = glob(["*.java"]),
)

java_library(
    name = "writer",
    srcs = [
        ":writer_files",
    ],
    deps = [
        "//src/main/java/com/google/api/generator/engine/ast",
        "//src/main/java/com/google/api/generator/engine/lexicon",
        "@com_google_guava_guava//jar",
        "@google_java_format_all_deps//jar",
    ],
)
