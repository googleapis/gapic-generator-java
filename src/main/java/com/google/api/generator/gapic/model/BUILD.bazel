load("@rules_java//java:defs.bzl", "java_library")

package(default_visibility = ["//visibility:public"])

filegroup(
    name = "model_files",
    srcs = glob(["*.java"]),
)

java_library(
    name = "model",
    srcs = [
        ":model_files",
    ],
    deps = [
        "//:service_config_java_proto",
        "//src/main/java/com/google/api/generator:autovalue",
        "//src/main/java/com/google/api/generator/engine/ast",
        "//src/main/java/com/google/api/generator/gapic/utils",
        "@com_google_api_api_common//jar",
        "@com_google_auto_value_auto_value//jar",
        "@com_google_auto_value_auto_value_annotations//jar",
        "@com_google_code_findbugs_jsr305//jar",
        "@com_google_googleapis//google/api:api_java_proto",
        "@com_google_googleapis//google/rpc:rpc_java_proto",
        "@com_google_guava_guava//jar",
        "@com_google_protobuf//:protobuf_java",
    ],
)
