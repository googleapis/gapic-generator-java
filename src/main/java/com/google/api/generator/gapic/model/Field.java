// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.api.generator.gapic.model;

import com.google.api.generator.engine.ast.TypeNode;
import com.google.auto.value.AutoValue;
import javax.annotation.Nullable;

@AutoValue
public abstract class Field {
  public abstract String name();

  public abstract TypeNode type();

  public abstract boolean isRepeated();

  @Nullable
  public abstract ResourceReference resourceReference();

  public boolean hasResourceReference() {
    return type().equals(TypeNode.STRING) && resourceReference() != null;
  }

  public static Builder builder() {
    return new AutoValue_Field.Builder().setIsRepeated(false);
  }

  @AutoValue.Builder
  public abstract static class Builder {
    public abstract Builder setName(String name);

    public abstract Builder setType(TypeNode type);

    public abstract Builder setIsRepeated(boolean isRepeated);

    public abstract Builder setResourceReference(ResourceReference resourceReference);

    public abstract Field build();
  }
}
