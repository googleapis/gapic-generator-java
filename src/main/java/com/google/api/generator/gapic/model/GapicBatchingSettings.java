// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.api.generator.gapic.model;

import com.google.auto.value.AutoValue;
import javax.annotation.Nullable;

@AutoValue
public abstract class GapicBatchingSettings {
  public enum FlowControlLimitExceededBehavior {
    THROW_EXCEPTION,
    BLOCK,
    IGNORE
  };

  public abstract String protoPakkage();

  public abstract String serviceName();

  public abstract String methodName();

  public abstract int elementCountThreshold();

  public abstract long requestByteThreshold();

  public abstract long delayThresholdMillis();

  @Nullable
  public abstract Integer flowControlElementLimit();

  @Nullable
  public abstract Integer flowControlByteLimit();

  public abstract FlowControlLimitExceededBehavior flowControlLimitExceededBehavior();

  public static Builder builder() {
    return new AutoValue_GapicBatchingSettings.Builder()
        .setFlowControlLimitExceededBehavior(FlowControlLimitExceededBehavior.IGNORE);
  }

  @AutoValue.Builder
  public abstract static class Builder {
    public abstract Builder setProtoPakkage(String protoPakkage);

    public abstract Builder setServiceName(String serviceName);

    public abstract Builder setMethodName(String methodName);

    public abstract Builder setElementCountThreshold(int elementCountThreshold);

    public abstract Builder setRequestByteThreshold(long requestByteThreshold);

    public abstract Builder setDelayThresholdMillis(long delalyThresholdMillis);

    public abstract Builder setFlowControlElementLimit(Integer flowControlElementLimit);

    public abstract Builder setFlowControlByteLimit(Integer flowControlByteLimit);

    public abstract Builder setFlowControlLimitExceededBehavior(
        FlowControlLimitExceededBehavior behavior);

    public abstract GapicBatchingSettings build();
  }
}
