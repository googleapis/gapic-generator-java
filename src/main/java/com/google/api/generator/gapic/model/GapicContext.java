// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.api.generator.gapic.model;

import com.google.auto.value.AutoValue;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import java.util.List;
import java.util.Map;

@AutoValue
public abstract class GapicContext {
  // Maps the message name (as it appears in the protobuf) to Messages.
  public abstract ImmutableMap<String, Message> messages();

  // Maps the resource type string to ResourceNames.
  public abstract ImmutableMap<String, ResourceName> resourceNames();

  public abstract ImmutableList<Service> services();

  public static Builder builder() {
    return new AutoValue_GapicContext.Builder();
  }

  @AutoValue.Builder
  public abstract static class Builder {
    public abstract Builder setMessages(Map<String, Message> messages);

    public abstract Builder setResourceNames(Map<String, ResourceName> resourceNames);

    public abstract Builder setServices(List<Service> services);

    public abstract GapicContext build();
  }
}
