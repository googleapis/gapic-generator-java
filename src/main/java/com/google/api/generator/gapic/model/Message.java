// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.api.generator.gapic.model;

import com.google.api.generator.engine.ast.TypeNode;
import com.google.auto.value.AutoValue;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@AutoValue
public abstract class Message {
  public abstract String name();

  // TODO(unsupported): oneof fields are parsed as separate ones because field flattening refers to
  // a specific field.
  public abstract ImmutableList<Field> fields();

  public abstract TypeNode type();

  public abstract ImmutableMap<String, Field> fieldMap();

  public static Builder builder() {
    return new AutoValue_Message.Builder();
  }

  @AutoValue.Builder
  public abstract static class Builder {
    public abstract Builder setName(String name);

    public abstract Builder setFields(List<Field> fields);

    public abstract Builder setType(TypeNode type);

    abstract Builder setFieldMap(Map<String, Field> fieldMap);

    abstract ImmutableList<Field> fields();

    abstract Message autoBuild();

    public Message build() {
      setFieldMap(fields().stream().collect(Collectors.toMap(f -> f.name(), f -> f)));
      return autoBuild();
    }
  }
}
