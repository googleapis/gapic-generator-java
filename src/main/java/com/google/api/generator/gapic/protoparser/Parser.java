// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.api.generator.gapic.protoparser;

import com.google.api.generator.engine.ast.TypeNode;
import com.google.api.generator.engine.ast.VaporReference;
import com.google.api.generator.gapic.model.Field;
import com.google.api.generator.gapic.model.Message;
import com.google.api.generator.gapic.model.Method;
import com.google.api.generator.gapic.model.Service;
import com.google.common.base.Preconditions;
import com.google.common.collect.Maps;
import com.google.protobuf.DescriptorProtos.FileDescriptorProto;
import com.google.protobuf.Descriptors.Descriptor;
import com.google.protobuf.Descriptors.DescriptorValidationException;
import com.google.protobuf.Descriptors.FileDescriptor;
import com.google.protobuf.Descriptors.MethodDescriptor;
import com.google.protobuf.Descriptors.ServiceDescriptor;
import com.google.protobuf.compiler.PluginProtos.CodeGeneratorRequest;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class Parser {
  static class GapicParserException extends RuntimeException {
    public GapicParserException(String errorMessage) {
      super(errorMessage);
    }
  }

  public static List<Service> parseServices(
      CodeGeneratorRequest request, Map<String, Message> messageTypes) {
    Map<String, FileDescriptor> fileDescriptors = getFilesToGenerate(request);
    List<Service> services = new ArrayList<>();
    for (String fileToGenerate : request.getFileToGenerateList()) {
      FileDescriptor fileDescriptor =
          Preconditions.checkNotNull(
              fileDescriptors.get(fileToGenerate),
              "Missing file descriptor for [%s]",
              fileToGenerate);

      String pakkage = TypeParser.getPackage(fileDescriptor);
      for (ServiceDescriptor serviceDescriptor : fileDescriptor.getServices()) {
        List<Method> methods = parseMethods(serviceDescriptor, messageTypes);
        Service service =
            Service.builder()
                .setName(serviceDescriptor.getName())
                .setPakkage(pakkage)
                .setMethods(methods)
                .build();
        services.add(service);
      }
    }

    return services;
  }

  public static Map<String, Message> parseMessages(CodeGeneratorRequest request) {
    Map<String, FileDescriptor> fileDescriptors = getFilesToGenerate(request);
    Map<String, Message> messages = new HashMap<>();
    for (String fileToGenerate : request.getFileToGenerateList()) {
      FileDescriptor fileDescriptor =
          Preconditions.checkNotNull(
              fileDescriptors.get(fileToGenerate),
              "Missing file descriptor for [%s]",
              fileToGenerate);

      String pakkage = TypeParser.getPackage(fileDescriptor);
      for (Descriptor messageDescriptor : fileDescriptor.getMessageTypes()) {
        List<Field> fields = parseFields(messageDescriptor);
        String messageName = messageDescriptor.getName();
        messages.put(
            messageName,
            Message.builder()
                .setType(
                    TypeNode.withReference(
                        VaporReference.builder().setName(messageName).setPakkage(pakkage).build()))
                .setName(messageName)
                .setFields(fields)
                .build());
      }
    }
    return messages;
  }

  private static List<Field> parseFields(Descriptor messageDescriptor) {
    return messageDescriptor.getFields().stream()
        .map(f -> Field.builder().setName(f.getName()).setType(TypeParser.parseType(f)).build())
        .collect(Collectors.toList());
  }

  private static Map<String, FileDescriptor> getFilesToGenerate(CodeGeneratorRequest request) {
    // Build the fileDescriptors map so that we can create the FDs for the filesToGenerate.
    Map<String, FileDescriptor> fileDescriptors = Maps.newHashMap();
    for (FileDescriptorProto fileDescriptorProto : request.getProtoFileList()) {
      // Look up the imported files from previous file descriptors.  It is sufficient to look at
      // only previous file descriptors because CodeGeneratorRequest guarantees that the files
      // are sorted in topological order.
      FileDescriptor[] deps = new FileDescriptor[fileDescriptorProto.getDependencyCount()];
      for (int i = 0; i < fileDescriptorProto.getDependencyCount(); i++) {
        String name = fileDescriptorProto.getDependency(i);
        deps[i] =
            Preconditions.checkNotNull(
                fileDescriptors.get(name), "Missing file descriptor for [%s]", name);
      }

      FileDescriptor fileDescriptor = null;
      try {
        fileDescriptor = FileDescriptor.buildFrom(fileDescriptorProto, deps);
      } catch (DescriptorValidationException e) {
        throw new GapicParserException(e.getMessage());
      }

      fileDescriptors.put(fileDescriptor.getName(), fileDescriptor);
    }
    return fileDescriptors;
  }

  private static List<Method> parseMethods(
      ServiceDescriptor serviceDescriptor, Map<String, Message> messageTypes) {
    List<Method> methods = new ArrayList<>();
    for (MethodDescriptor methodDescriptor : serviceDescriptor.getMethods()) {
      methods.add(
          Method.builder()
              .setName(methodDescriptor.getName())
              .setInputType(TypeParser.parseType(methodDescriptor.getInputType()))
              .setOutputType(TypeParser.parseType(methodDescriptor.getOutputType()))
              .build());
    }
    return methods;
  }
}
