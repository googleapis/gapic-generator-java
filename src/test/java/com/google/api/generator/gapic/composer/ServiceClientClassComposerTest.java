// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.api.generator.gapic.composer;

import static junit.framework.Assert.assertEquals;

import com.google.api.generator.engine.writer.JavaWriterVisitor;
import com.google.api.generator.gapic.model.GapicClass;
import com.google.api.generator.gapic.model.Message;
import com.google.api.generator.gapic.model.Service;
import com.google.api.generator.gapic.protoparser.Parser;
import com.google.protobuf.Descriptors.FileDescriptor;
import com.google.protobuf.Descriptors.ServiceDescriptor;
import com.google.showcase.v1beta1.EchoOuterClass;
import java.util.List;
import java.util.Map;
import org.junit.Before;
import org.junit.Test;

public class ServiceClientClassComposerTest {
  private ServiceDescriptor echoService;
  private FileDescriptor echoFileDescriptor;

  @Before
  public void setUp() {
    echoFileDescriptor = EchoOuterClass.getDescriptor();
    echoService = echoFileDescriptor.getServices().get(0);
    assertEquals(echoService.getName(), "Echo");
  }

  @Test
  public void generateServiceClasses() {
    Map<String, Message> messageTypes = Parser.parseMessages(echoFileDescriptor);
    List<Service> services = Parser.parseService(echoFileDescriptor, messageTypes);
    Service echoProtoService = services.get(0);
    GapicClass clazz =
        ServiceClientClassComposer.instance().generate(echoProtoService, messageTypes);

    JavaWriterVisitor visitor = new JavaWriterVisitor();
    clazz.classDefinition().accept(visitor);
    assertEquals(EXPECTED_CLASS_STRING, visitor.write());
  }

  // TODO(miraleung): Update this when a file-diffing test mechanism is in place.
  private static final String EXPECTED_CLASS_STRING =
      "package com.google.showcase.v1beta1;\n"
          + "\n"
          + "import com.google.api.gax.core.BackgroundResource;\n"
          + "import com.google.api.gax.longrunning.OperationFuture;\n"
          + "import com.google.api.gax.rpc.BidiStreamingCallable;\n"
          + "import com.google.api.gax.rpc.ClientStreamingCallable;\n"
          + "import com.google.api.gax.rpc.OperationCallable;\n"
          + "import com.google.api.gax.rpc.ServerStreamingCallable;\n"
          + "import com.google.api.gax.rpc.UnaryCallable;\n"
          + "import com.google.longrunning.Operation;\n"
          + "import com.google.longrunning.OperationsClient;\n"
          + "import com.google.rpc.Status;\n"
          + "import com.google.showcase.v1beta1.stub.EchoStub;\n"
          + "import java.io.IOException;\n"
          + "import java.util.concurrent.TimeUnit;\n"
          + "import javax.annotation.Generated;\n"
          + "\n"
          + "@Generated(\"by gapic-generator\")\n"
          + "public class EchoClient implements BackgroundResource {\n"
          + "  private final EchoSettings settings;\n"
          + "  private final EchoStub stub;\n"
          + "  private final OperationsClient operationsClient;\n"
          + "\n"
          + "  public static final EchoClient create(EchoSettings settings) throws IOException {\n"
          + "    return create(EchoSettings.newBuilder().builder());\n"
          + "  }\n"
          + "\n"
          + "  public EchoStub getStub() {\n"
          + "    return stub;\n"
          + "  }\n"
          + "\n"
          + "  public final EchoSettings getSettings() {\n"
          + "    return settings;\n"
          + "  }\n"
          + "\n"
          + "  public final OperationsClient getOperationsClient() {\n"
          + "    return operationsClient;\n"
          + "  }\n"
          + "\n"
          + "  public final EchoResponse echo(String content) {\n"
          + "    EchoRequest request = EchoRequest.newBuilder().setContent(content).build();\n"
          + "    return echo(request);\n"
          + "  }\n"
          + "\n"
          + "  public final EchoResponse echo(Status error) {\n"
          + "    EchoRequest request = EchoRequest.newBuilder().setError(error).build();\n"
          + "    return echo(request);\n"
          + "  }\n"
          + "\n"
          + "  public final EchoResponse echo(String content, Severity severity) {\n"
          + "    EchoRequest request =\n"
          + "        EchoRequest.newBuilder().setContent(content).setSeverity(severity).build();\n"
          + "    return echo(request);\n"
          + "  }\n"
          + "\n"
          + "  public final EchoResponse echo(EchoRequest request) {\n"
          + "    return echoCallable().call(request);\n"
          + "  }\n"
          + "\n"
          + "  public final UnaryCallable<EchoRequest, EchoResponse> echoCallable() {\n"
          + "    return stub.echoCallable();\n"
          + "  }\n"
          + "\n"
          + "  public final ServerStreamingCallable<ExpandRequest, EchoResponse> expandCallable()"
          + " {\n"
          + "    return stub.expandCallable();\n"
          + "  }\n"
          + "\n"
          + "  public final ClientStreamingCallable<EchoRequest, EchoResponse> collectCallable()"
          + " {\n"
          + "    return stub.collectCallable();\n"
          + "  }\n"
          + "\n"
          + "  public final BidiStreamingCallable<EchoRequest, EchoResponse> chatCallable() {\n"
          + "    return stub.chatCallable();\n"
          + "  }\n"
          + "\n"
          + "  public final BidiStreamingCallable<EchoRequest, EchoResponse> chatAgainCallable()"
          + " {\n"
          + "    return stub.chatAgainCallable();\n"
          + "  }\n"
          + "\n"
          + "  public final PagedExpandResponse pagedExpand(PagedExpandRequest request) {\n"
          + "    return pagedExpandCallable().call(request);\n"
          + "  }\n"
          + "\n"
          + "  public final UnaryCallable<PagedExpandRequest, PagedExpandResponse>"
          + " pagedExpandCallable() {\n"
          + "    return stub.pagedExpandCallable();\n"
          + "  }\n"
          + "\n"
          + "  public final OperationFuture<WaitResponse, WaitMetadata> waitAsync(WaitRequest"
          + " request) {\n"
          + "    return waitOperationCallable().futureCall(request);\n"
          + "  }\n"
          + "\n"
          + "  public final OperationCallable<WaitRequest, WaitResponse, WaitMetadata>"
          + " waitOperationCallabke() {\n"
          + "    return stub.waitOperationCallabke();\n"
          + "  }\n"
          + "\n"
          + "  public final UnaryCallable<WaitRequest, Operation> waitCallable() {\n"
          + "    return stub.waitCallable();\n"
          + "  }\n"
          + "\n"
          + "  public final BlockResponse block(BlockRequest request) {\n"
          + "    return blockCallable().call(request);\n"
          + "  }\n"
          + "\n"
          + "  public final UnaryCallable<BlockRequest, BlockResponse> blockCallable() {\n"
          + "    return stub.blockCallable();\n"
          + "  }\n"
          + "\n"
          + "  @Override\n"
          + "  public void close() {}\n"
          + "\n"
          + "  @Override\n"
          + "  public void shutdown() {}\n"
          + "\n"
          + "  @Override\n"
          + "  public boolean isShutdown() {\n"
          + "    return false;\n"
          + "  }\n"
          + "\n"
          + "  @Override\n"
          + "  public boolean isTerminated() {\n"
          + "    return false;\n"
          + "  }\n"
          + "\n"
          + "  @Override\n"
          + "  public void shutdownNow() {}\n"
          + "\n"
          + "  @Override\n"
          + "  public boolean awaitTermination(long duration, TimeUnit unit) throws"
          + " InterruptedException {\n"
          + "    return false;\n"
          + "  }\n"
          + "}\n";
}
