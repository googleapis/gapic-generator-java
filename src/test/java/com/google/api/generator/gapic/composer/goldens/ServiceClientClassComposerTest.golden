package com.google.showcase.v1beta1;

import com.google.api.core.BetaApi;
import com.google.api.gax.core.BackgroundResource;
import com.google.api.gax.longrunning.OperationFuture;
import com.google.api.gax.rpc.BidiStreamingCallable;
import com.google.api.gax.rpc.ClientStreamingCallable;
import com.google.api.gax.rpc.OperationCallable;
import com.google.api.gax.rpc.ServerStreamingCallable;
import com.google.api.gax.rpc.UnaryCallable;
import com.google.api.resourcenames.ResourceName;
import com.google.common.base.Strings;
import com.google.longrunning.Operation;
import com.google.longrunning.OperationsClient;
import com.google.rpc.Status;
import com.google.showcase.v1beta1.stub.EchoStub;
import com.google.showcase.v1beta1.stub.EchoStubSettings;
import java.io.IOException;
import java.util.concurrent.TimeUnit;
import javax.annotation.Generated;

@BetaApi
@Generated("by gapic-generator")
public class EchoClient implements BackgroundResource {
  private final EchoSettings settings;
  private final EchoStub stub;
  private final OperationsClient operationsClient;

  public static final EchoClient create() throws IOException {
    return create(EchoSettings.newBuilder().build());
  }

  public static final EchoClient create(EchoSettings settings) throws IOException {
    return new EchoClient(settings);
  }

  @BetaApi("A restructuring of stub classes is planned, so this may break in the future")
  public static final EchoClient create(EchoStub stub) {
    return new EchoClient(stub);
  }

  protected EchoClient(EchoSettings settings) throws IOException {
    this.settings = settings;
    this.stub = ((EchoStubSettings) settings.getStubSettings()).createStub();
    this.operationsClient = OperationsClient.create(this.stub.getOperationsStub());
  }

  @BetaApi("A restructuring of stub classes is planned, so this may break in the future")
  protected EchoClient(EchoStub stub) {
    this.settings = null;
    this.stub = stub;
    this.operationsClient = OperationsClient.create(this.stub.getOperationsStub());
  }

  public final EchoSettings getSettings() {
    return settings;
  }

  @BetaApi("A restructuring of stub classes is planned, so this may break in the future")
  public EchoStub getStub() {
    return stub;
  }

  public final OperationsClient getOperationsClient() {
    return operationsClient;
  }

  public final EchoResponse echo(String content) {
    EchoRequest request = EchoRequest.newBuilder().setContent(content).build();
    return echo(request);
  }

  public final EchoResponse echo(Status error) {
    EchoRequest request = EchoRequest.newBuilder().setError(error).build();
    return echo(request);
  }

  public final EchoResponse echo(String content, Severity severity) {
    EchoRequest request =
        EchoRequest.newBuilder().setContent(content).setSeverity(severity).build();
    return echo(request);
  }

  public final EchoResponse echo(String name) {
    EchoRequest request = EchoRequest.newBuilder().setName(name).build();
    return echo(request);
  }

  public final EchoResponse echo(FoobarName name) {
    EchoRequest request =
        EchoRequest.newBuilder()
            .setName(Strings.isNullOrEmpty(name) ? null : name.toString())
            .build();
    return echo(request);
  }

  public final EchoResponse echo(String parent) {
    EchoRequest request = EchoRequest.newBuilder().setParent(parent).build();
    return echo(request);
  }

  public final EchoResponse echo(ResourceName parent) {
    EchoRequest request =
        EchoRequest.newBuilder()
            .setParent(Strings.isNullOrEmpty(parent) ? null : parent.toString())
            .build();
    return echo(request);
  }

  public final EchoResponse echo(EchoRequest request) {
    return echoCallable().call(request);
  }

  public final UnaryCallable<EchoRequest, EchoResponse> echoCallable() {
    return stub.echoCallable();
  }

  public final ServerStreamingCallable<ExpandRequest, EchoResponse> expandCallable() {
    return stub.expandCallable();
  }

  public final ClientStreamingCallable<EchoRequest, EchoResponse> collectCallable() {
    return stub.collectCallable();
  }

  public final BidiStreamingCallable<EchoRequest, EchoResponse> chatCallable() {
    return stub.chatCallable();
  }

  public final BidiStreamingCallable<EchoRequest, EchoResponse> chatAgainCallable() {
    return stub.chatAgainCallable();
  }

  public final PagedExpandResponse pagedExpand(PagedExpandRequest request) {
    return pagedExpandCallable().call(request);
  }

  public final UnaryCallable<PagedExpandRequest, PagedExpandResponse> pagedExpandCallable() {
    return stub.pagedExpandCallable();
  }

  public final OperationFuture<WaitResponse, WaitMetadata> waitAsync(WaitRequest request) {
    return waitOperationCallable().futureCall(request);
  }

  public final OperationCallable<WaitRequest, WaitResponse, WaitMetadata> waitOperationCallabke() {
    return stub.waitOperationCallabke();
  }

  public final UnaryCallable<WaitRequest, Operation> waitCallable() {
    return stub.waitCallable();
  }

  public final BlockResponse block(BlockRequest request) {
    return blockCallable().call(request);
  }

  public final UnaryCallable<BlockRequest, BlockResponse> blockCallable() {
    return stub.blockCallable();
  }

  @Override
  public final void close() {
    stub.close();
  }

  @Override
  public void shutdown() {
    stub.shutdown();
  }

  @Override
  public boolean isShutdown() {
    return stub.isShutdown();
  }

  @Override
  public boolean isTerminated() {
    return stub.isTerminated();
  }

  @Override
  public void shutdownNow() {
    stub.shutdownNow();
  }

  @Override
  public boolean awaitTermination(long duration, TimeUnit unit) throws InterruptedException {
    return stub.awaitTermination(duration, unit);
  }
}
